src/simulations/collect.jl

```diff
@@ -0,0 +1,361 @@
+###################################################
+# Definition of the data collection API
+###################################################
+get_data(a, s::Symbol) = getproperty(a, s)
+get_data(a, f::Function) = f(a)
+
+# TODO: `add_data` is a misleading name, we need a better one.
+add_data(s, model, when::AbstractVector) = s ∈ when
+add_data(s, model, when::Bool) = when
+add_data(s, model, when) = when(model, s)
+
+"""
+    run!(model, agent_step! [, model_step!], n; kwargs...) → agent_df, model_df
+
+Run the model (step it with the input arguments propagated into `step!`) and collect
+data specified by the keywords, explained one by one below. Return the data as
+two `DataFrame`s, one for agent-level data and one for model-level data.
+
+## Data-deciding keywords
+`agent_properties::Vector` decides the agent data. If an entry is a `name::Symbol`,
+then the data for this entry is agent's field with that `name`. If an entry is a `f::Function`,
+then the data for this entry is just `f(a)` for each agent `a`.
+
+`model_properties::Vector` works exactly like `agent_properties` but for model level data.
+
+`aggregation_dict` decides whether the agent data should be aggregated over agents.
+
+#TODO: Describe what `aggregation_dict` is.
+
+# TODO: Well, how do you use aggregation on entries of `agent_properties` that are
+**functions**? Maybe it is better to make `aggregation_dict` keys to be integers,
+that are the entries of `agent_properties`, instead of `symbols` that are only limited
+to agent fields and not functions.
+
+# TODO: isn't it simpler to make `aggregation_dict` a VECTOR just like
```

Comment:
A dict is just easier to write and read, especially when you want to apply multiple functions to each field.

Reply:
good point. So how do we handle cases where the collected data are functions of agents instead of just fields?

Reply:
Wait wait. Aggregation happens on column _names_ so they can use the name of `f` ? like `:f` ?

Reply:
That's why I am asking you to *explain* what `aggregation_dict` is. Is it exactly the same thing as before? Just write it out explicitly, and then I don't have to assume any more...